doc/user_guide/introduction_to_psykal.rst

```diff
@@ -0,0 +1,408 @@
+the algorithm layer but do not require an associated kernel implementation
+because they are provided by the infrastructure.
+
+These are useful for commonly used operations, as they reduce the amount of
+code that the PSyKAl project has to maintain. In addition, they offer
+potential performance advantages as their implementation can completely
+change for different architectures and the PSy layer is free to implement
+these operations in whatever way it chooses.
+
+The list of supported Built-ins is API-specific and therefore these are
+described under the documentation of each API. In general, PSyclone will
+need to know the types of the arguments being passed to any Built-ins.
+Each API provides a Fortran file that contains the metadata for all
+Built-in operations supported for that API.
+
+.. note:: When a particular Built-in is used, the name of this
+          Built-in should not be used for anything else within the
+          same scope. For example, it is not valid to make use of a
+          Built-in called ``setval_c`` and for its parent subroutine
+          to also be called ``setval_c``. In this case PSyclone will
+          raise an exception.
+
+Example
++++++++
+
+In the following algorithm-layer example from the LFRic API, the invoke call includes a call
+to two Built-ins (``setval_c`` and ``X_divideby_Y``) and a user-supplied kernel
+(``matrix_vector_kernel_mm_type``).
+The ``setval_c`` Built-in sets all values in the field ``Ax`` to ``1.0`` and
+the ``X_divideby_Y`` Built-in divides values in the field ``rhs`` by their
+equivalent (per degree of freedom) values in the field ``lumped_weight``
+(see :ref:`supported LFRic API Built-ins <lfric-built-ins>`).
+Notice that, unlike the kernel call, no ``use`` association is required for
+the Built-ins since they are provided as part of the environment.
+::
+
+  module solver_mod
+    ...
+    use matrix_vector_mm_mod, only: matrix_vector_kernel_mm_type
+    ...
+
+    subroutine jacobi_solver_algorithm(lhs, rhs, mm, mesh, n_iter)
+
+      integer(kind=i_def), intent(in)    :: n_iter
+      type(field_type),    intent(inout) :: lhs
+      type(field_type),    intent(in)    :: rhs
+      type(operator_type), intent(in)    :: mm
+      type(mesh_type),     intent(in)    :: mesh
+      type(field_type)                   :: Ax, lumped_weight, res
+
+      real(kind=r_def), parameter :: MU = 0.9_r_def
+      ...
+
+      ! Compute mass lump
+      call invoke( name = "Jacobi_mass_lump",                           &
+                   setval_c(Ax, 1.0_r_def),                             &
+                   matrix_vector_kernel_mm_type(lumped_weight, Ax, mm), &
+                   X_divideby_Y(lhs, rhs, lumped_weight) )
+
+    end subroutine jacobi_solver_algorithm
+    ...
+
+  end module solver_mod
+
+
+.. _PSy-layer:
+
+PSy layer
+---------
+
+In the PSyKAl model, the PSy layer is the bridge between the Algorithm
+full field operations and the Kernel and Built-Ins individual element
```

Comment:
"full-field"

Reply:
and "Kernel/Built-Ins individual..."

Reply:
Done

Reply:
Almost, should be "full-field", not "full-fields".